### ISSUE TITLE
HTTP caching

### ISSUE THREAD

**Issue by rubensworks** [comunica/comunica#166]
#### Issue type:

- :heavy_plus_sign: Feature request

____
#### Description:

We should add support for proper client-side caching of HTTP resources.
We should check both the `Expires`, `Cache-Control`, `ETag` and `Last-Modified` headers, and appropriately send requests with `If-None-Match`/`If-Modified-Since` if needed.

We should make the cache actor work on the http bus.

**Comment by pietercolpaert** [comment]
I believe we should rely on https://www.npmjs.com/package/http-cache-semantics

The cache probably should not be mediated, but it should be statically configured: e.g., if you have a redis cache, you will need a config file to define how to connect to this, or maybe you’d rather want to use something like [LevelDB](https://github.com/Level/levelup) or [cacache](https://www.npmjs.com/package/cacache). As long as somethin adheres to https://developer.mozilla.org/en-US/docs/Web/API/Cache, I would allow it through a kind of strategy pattern that reads a config. What do you think?

**Comment by rubensworks** [comment]
That package looks like it might do the trick.

> The cache probably should not be mediated, but it should be statically configured

If with "not mediated", you mean that it should not be on a separate "caching" bus, then I agree :-) But it should still listen on the HTTP bus IMO.

We could have this caching actor observe all HTTP calls passively (which will cache all performed HTTP requests), and also actively be registered on the HTTP bus (for replying with results from cache).

> a kind of strategy pattern that reads a config

Yes, that sounds good. The caching storage strategy (redis, LevelDB, ...) could be defined in the config file via a parameter. We'd then need separate _caching storage_ components (not necessarily actors (I think)) as well.

**Comment by rubensworks** [comment]
To consider: https://github.com/solid/web-access-control-spec/pull/67

**Comment by jaxoncreed** [comment]
As discussed with @rubensworks, I will work on this issue via the Comunica Association.

**Comment by jaxoncreed** [comment]
I've encountered a few problems with this task:

> A simple default LRUCache-based Cache implementation should be set here if the user provided none in the context: https://github.com/comunica/comunica/blob/master/packages/actor-init-query/lib/QueryEngineBase.ts#L110
It can be implemented using https://www.npmjs.com/package/http-cache-semantics

The `http-cache-semantics` library is primarily concerned with the "cache-control" header and the web-cache interface doesn't care about that (for example see https://w3c.github.io/ServiceWorker/#cache-addAll the only thing it cares about is the status of the response and whether or not its http)

Instead of using the `http-cache-semantics` library, I could just program the Cache interface the way the specification tells me to, but I don't think that's what we're going for.

Maybe abandoning the goal to build a node version of the Cache interface is the best way to proceed. Instead we can build a cache around http-cache-semantics that doesn't follow the web-cache interface.

Also neither `http-cache-semantics` nor the Cache interface specification touch `E-Tag` and `If-None-Match`, two things that I think our caching solution should consider.

So, perhaps the best solution is to write a generic interface that does consider these things. I could keep with the Web-Cache interface, but in that case, we shouldn't call it a "Web-Cache" since it doesn't follow the spec. Maybe we could call it `http-cache-lru`

**Comment by rubensworks** [comment]
Ok, let's get rid of `add` and `addAll`. But I still think reusing (part of) the Cache interface is valuable, even if we only implement parts of it.

My suggestion is to only implement `match`, (perhaps `matchAll`?) `put`, `delete`, and `keys`, and throw a `NotImplementedError` on all other methods.
And we don't have to follow the exact steps that are prescribed in [the Web Cache spec](https://w3c.github.io/ServiceWorker/#cache-interface).
Within Comunica, this cache would probably only even use `match` (for reading from cache), `put` (for caching requests that have been done by another HTTP actor) and `delete` (for manually invalidating requests, possibly with the help of `keys`).

So the general flow of the caching actor would be:
```
const cachedResponse = await cache.match(originalRequestAction);
if (cachedResponse)
  return cachedResponse;

const response = await mediatorHttp.mediate(originalRequestAction);
const [ clonedResponse1, clonedResponse2 ] = someMethodToCloneResponse(response);
await cache.put(originalRequestAction, cloneResponse1);
return clonedResponse2;
```

Within the `match` and `put`, we would still be free to implement support E-Tags and anything we like.

**Comment by jaxoncreed** [comment]
Okay awesome, @rubensworks!  That was easy. I've completed the library here: https://github.com/o-development/web-cache-lru/settings. What are the next steps to getting it under Comunica umbrella?

**Comment by rubensworks** [comment]
@jaxoncreed Nice!

> What are the next steps to getting it under Comunica umbrella?

I would suggest just putting it inside the Comunica monorepo as a new package under `packages/`.
This would make it easier to tweak things to it in case we notice new requirements when using it from within `@comunica/actor-http-cache`.
So starting your own fork, and creating a PR with it in it might the the easiest path forward.

**Comment by jaxoncreed** [comment]
Sorry to keep coming back to this point. But I think there's a problem with this:

> A simple default LRUCache-based Cache implementation should be set here if the user provided none in the context: 

I don't think we should be asking users to provide a Cache in context if our library is expecting a Cache that doesn't perform in a spec compliant way.

Instead I think we should have a simple key value interface that gets set in context. Then we'll have comunica take care of the cache semantics. That way you could pass in a redis, or localstore, or any kind of key value interface as a cache without having to worry about re-implementing cache semantics.

**Comment by jaxoncreed** [comment]
I implemented the first suggestion, but now I think we shouldn't be aligned with Web Cache interface at all. What's the point of aligning with something at an interface level if you're not going to implement it at a spec level? If someone passes in a spec-compliant Cache, comunica will break because the spec-compliant Cache doesn't care about http cache semantics.

By allowing developers to pass in a WebCache, we are communicating to them that it's fine of pass in a spec-compliant WebCache, when it is not fine. It's confusing to developers and causes code like this:

![image](https://user-images.githubusercontent.com/5464743/172916436-3e56f5cf-672a-45a3-8539-eda071a03198.png)

I really think the best way forward is to not use the `Cache` interface at all. The `CacheStore` should just be a simple key value store.

**Comment by rubensworks** [comment]
> The CacheStore should just be a simple key value store.

Ok, go ahead!

Will you also be decoupling the cache store from the cache semantics?

**Comment by jaxoncreed** [comment]
Yep, I think decoupling them makes sense.

**Comment by rubensworks** [comment]
@jaxoncreed I'll unassign you from this bounty. Please let me know if you'd like to come back to it.

**Comment by jaxoncreed** [comment]
Thanks, sorry this was never completed.

**Comment by rubensworks** [comment]
Great to hear you're interested in this @iMac7! Could you contact me via ruben.taelman@ugent.be ?

**Comment by jeswr** [comment]
Note there is also work taking place to implement HTTP caching directly in the node fetch function https://github.com/nodejs/undici/issues/3231. In particular [rfc9111](https://www.rfc-editor.org/rfc/rfc9111.html) is being implemented as part of that work - so it may be worth coordinating any work that overlaps :)

**Comment by rubensworks** [comment]
Good idea, thanks for the input @jeswr!

@iMac7, it may indeed be a good idea to check with @Uzlopak to see if there's overlapping work.
I would expect the caching of Response object to at least be reusable somehow.

**Comment by jeswr** [comment]
I only have as much information as is in that issue - I think a PR is yet to be opened as I only opened the issue a week ago.

**Comment by RubenEschauzier** [comment]
As discussed, I'll be working on this issue!

**Comment by jeswr** [comment]
FWIW I saw a PR go past implementing HTTP caching on the node-native fetch function. Might be worth re-using some of their code.

Most info can be discovered from https://github.com/nodejs/undici/issues/3231

**Comment by RubenEschauzier** [comment]
Good point, I checked it out; they avoid using the HTTP-cache-semantics package and instead do validation themselves. I don't think that is required for this PR right? In any case, re-using parts of their cache class is interesting, as I'm not 100% sure what cache store (like the `Cache` interface) to use.
